src/components/AudioPlayer/HighlightedLocationUpdater.tsx

```diff
@@ -0,0 +1,57 @@
+import { useEffect, useRef } from 'react';
+
+import { shallowEqual, useDispatch } from 'react-redux';
+import useSWRImmutable from 'swr/immutable';
+
+import getHighlightedLocation from './getHighlightedLocation';
+
+import { getChapterAudioFile } from 'src/api';
+import {
+  initialState as defaultHighlightedLocation,
+  setHighlightedLocation,
+} from 'src/redux/slices/QuranReader/highlightedLocation';
+import { makeChapterAudioFiles } from 'src/utils/apiPaths';
+
+type AudioTimestampsHighlightListenerProps = {
+  reciterId: number;
+  chapterId: number;
+  currentTime: number;
+};
+
+/**
+ * Listen to audio player's currentTime and update highlightedLocation state
+ *
+ * @returns {void}
+ */
+const HighlightedLocationUpdater = ({
+  reciterId,
+  chapterId,
+  currentTime,
+}: AudioTimestampsHighlightListenerProps) => {
+  const dispatch = useDispatch();
+  const { data } = useSWRImmutable(makeChapterAudioFiles(reciterId, chapterId, true), () =>
+    getChapterAudioFile(reciterId, chapterId, true),
+  );
+
+  const lastHighlightedLocation = useRef(defaultHighlightedLocation);
+
+  // eslint-disable-next-line consistent-return
+  useEffect(() => {
+    if (!data) return;
+
+    const highlightedLocation = getHighlightedLocation(
```

Comment:
I think this still has room for performance improvement, we can do something like this

```javascript
// eslint-disable-next-line consistent-return
  useEffect(() => {
    if (!data) return;
    const currentTimeInMs = currentTime * 1000;

    let currentHighlightedVerse;
    if (
      lastHighlightedVerse.current &&
      currentTimeInMs >= lastHighlightedVerse.current.timestampFrom &&
      currentTimeInMs <= lastHighlightedVerse.current.timestampTo
    ) {
      currentHighlightedVerse = lastHighlightedVerse.current;
    } else {
      currentHighlightedVerse = data.verseTimings.find(
        (verse) => currentTimeInMs >= verse.timestampFrom && currentTimeInMs <= verse.timestampTo,
      );
      lastHighlightedVerse.current = currentHighlightedVerse;
    }

    const highlightedWordLocation = getHighlightedWordLocation(
      currentTimeInMs,
      currentHighlightedVerse,
    );
```

Reply:
so, it only do verseTimings.find when the condition doesn't match.

Reply:
we can also have `lastHighlightedWordLocation`

Reply:
+1, I actually have a similar proposal. Mentioning it in the loom.